rust/src/nasl/builtin/ssh/libssh/mod.rs

```diff
@@ -0,0 +1,152 @@
+// SPDX-FileCopyrightText: 2023 Greenbone AG
+//
+// SPDX-License-Identifier: GPL-2.0-or-later WITH x11vnc-openssl-exception
+
+//! Defines NASL ssh and sftp functions
+
+mod channel;
+mod session;
+
+use std::net::{IpAddr, UdpSocket};
+use std::{os::fd::AsRawFd, time::Duration};
+
+use libssh_rs::{LogLevel, SshOption};
+use russh::cipher;
+use russh_keys::key;
+use tokio::sync::{Mutex, MutexGuard};
+use tracing::debug;
+
+use super::error::Result;
+use super::Ssh;
+
+pub use libssh_rs::AuthMethods;
+pub use session::SshSession;
+
+pub type SessionId = i32;
+pub type Socket = std::os::raw::c_int;
+
+pub fn get_log_level() -> LogLevel {
+    let verbose = std::env::var("OPENVAS_LIBSSH_DEBUG")
+        .map(|x| x.parse::<i32>().unwrap_or_default())
+        .unwrap_or(0);
+
+    match verbose {
+        0 => LogLevel::NoLogging,
+        1 => LogLevel::Warning,
+        2 => LogLevel::Protocol,
+        3 => LogLevel::Packet,
+        _ => LogLevel::Functions,
+    }
+}
+
+fn to_comma_separated_string<T: AsRef<str>>(items: &[T]) -> Option<String> {
+    if items.is_empty() {
+        None
+    } else {
+        Some(
+            items
+                .iter()
+                .map(|name| name.as_ref().to_string())
+                .collect::<Vec<_>>()
+                .join(","),
+        )
+    }
+}
+
+type BorrowedSession<'a> = MutexGuard<'a, SshSession>;
+
+impl Ssh {
+    pub async fn find_id<'a>(
+        &'a self,
+        f: impl for<'b> Fn(&BorrowedSession<'b>) -> bool,
+    ) -> Result<Option<SessionId>> {
+        for id in self.ids() {
+            let session = self.get_by_id(*id).await?;
+            if f(&session) {
+                return Ok(Some(session.id()));
+            }
+        }
+        Ok(None)
+    }
+
+    /// Create a new session, but only add it to the list of active sessions
+    /// if the given closure which modifies the session returns Ok(...).
+    pub async fn add_new_session(
+        &mut self,
+        f: impl Fn(&mut BorrowedSession) -> Result<()>,
+    ) -> Result<SessionId> {
+        let id = self.next_session_id()?;
+        let session = Mutex::new(SshSession::new(id)?);
+        {
+            let mut borrowed_session = session.lock().await;
+            if let Err(e) = f(&mut borrowed_session) {
+                borrowed_session.disconnect().await?;
```

Comment:
q: I don't understand why you need to disconnect immediately.

Reply:
I did this to replicate the original code: libssh requires you to manually call the `disconnect` method if the session is dropped (as opposed to russh where RAII does its job). 
So the `add_new_session` method is effectively just making sure that the connection is closed if any error occurs during initialization (which effectively means that we will disconnect the session if any `set_option` or the `connect` call fails).

However, I now realize that this method is a little overkill - in the original implementation it was called from multiple places but I refactored it so that only `connect` calls this method anyways, so I should probably simplify it by replacing the closure with the code within `connect`, which should make it a little easier to understand.